README
Name
    Class::Usul::Build - Module::Build methods for standalone applications

Version
    This document describes Class::Usul::Build version v0.23.$Rev: 1 $

Synopsis
       use Class::Usul::Build;
       use MRO::Compat;

       my $builder = q(Class::Usul::Build);
       my $class   = $builder->subclass( class => 'Bob', code  => <<'EOB' );

       sub ACTION_instal { # Spelling mistake intentional
          my $self = shift;

          $self->next::method();

          # Your application specific post installation code goes here

          return;
       }
       EOB

Description
    Subclasses Module::Build. Ask questions during the install phase. The
    answers to the questions determine where the application will be
    installed and which additional actions will take place. Should be
    generic enough for any web application

ACTIONS
  ACTION_distmeta
  distmeta
    Updates license file and changelog

  ACTION_install
  install
    When called from it's subclass this method performs the sequence of
    actions required to install the application. Configuration options are
    written from the file build.json. The "actions" method returns the list
    of steps required to install the application

  ACTION_install_local_cpanm
  install_local_cpanm
    Install App::Cpanminus to the local lib

  ACTION_install_local_deps
  install_local_deps
    Installs dependencies to the local lib

  ACTION_install_local_lib
  install_local_lib
    Install local::lib locally

  ACTION_install_local_perl
  install_local_perl
    Install a specific Perl version to the local Perlbrew area

  ACTION_install_local_perlbrew
  install_local_perlbrew
    Installs Perlbrew locally

  ACTION_installdeps
  installdeps
    Iterates over the *requires* attributes calling CPAN each time to
    install the dependent module

  ACTION_local_archive
  local_archive
    Creates a tarball of the local lib directory

  ACTION_restore_local_archive
  restore_local_archive
    Unpacks an archive tarball of the local lib directory

  ACTION_standalone
  standalone
    Locally installs local lib and all dependencies

Subroutines/Methods
  class_path
       $path = $builder->class_path( $class_name );

    Returns the relative path to the specified class

  cli
       $cli = $builder->cli;

    Returns an instance of Class::Usul::Programs, the command line interface
    object

  cli_info
       $builder->cli_info( @list_of_messages );

    Calls info on the client object

  dispatch
    Overloads the M::B method. Calls "_setup_plugins" then the parent method

  distname
    Turns a class name into a distribution name

  install_actions_class
    Returns the class name of the class which contains the additional
    actions that are performed when the application is installed

  make_tarball
    Overloads the M::B method. Changes the directory which will contain the
    distribution tarball then calls the parent method

  patch_file
    Runs the *patch* utility on the specified source file

  post_install
       $builder->post_install( $config );

    Executes the custom post installation commands

  process_files
       $builder->process_files( $source, $destination );

    Handles the processing of files other than library modules and programs.
    Uses the *Bob::skip_pattern* defined in the subclass to select only
    those files that should be processed. Copies files from source to
    destination, creating the destination directories as required. Source
    can be a single file or a directory. The destination is optional and
    defaults to blib

  process_local_files
    Calls "process_file" setting the source to *local*

  question_class
    Returns the class name of the class which contains the questions that
    are asked when the application is installed

  set_base_path
       $base = $builder->set_base_path( $config );

    Uses the "$config->{style}" attribute to set the Module::Build
    *install_base* attribute to the base directory for this installation.
    Returns that path. Also sets; bin, lib, and var directory paths as
    appropriate. Called from ACTION_install

  skip_pattern
       $regexp = $builder->skip_pattern( $new_regexp );

    Accessor/mutator method. Used by "_copy_file" to skip processing files
    that match this pattern. Set to false to not have a skip list

  update_changelog
    Update the version number and date/time stamp in the Changes file

Private Methods
  _copy_file
       $builder->_copy_file( $source, $destination );

    Called by "process_files". Copies the $source file to the $destination
    directory

Diagnostics
    None

Configuration and Environment
    Edits and stores config information in the file build.json

Dependencies
    Class::Usul
    Module::Build
    Module::Metadata
    MRO::Compat
    Perl::Version
    Try::Tiny

Incompatibilities
    There are no known incompatibilities in this module

Bugs and Limitations
    There are no known bugs in this module. Please report problems to the
    address below. Patches are welcome

Author
    Peter Flanigan, "<Support at RoxSoft.co.uk>"

License and Copyright
    Copyright (c) 2013 Peter Flanigan. All rights reserved

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. See perlartistic

    This program is distributed in the hope that it will be useful, but
    WITHOUT WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE



Changes
Revision history for Class::Usul::Build

0.23.1    2013-08-13 13:21:07
          - Switched to Dist::Zilla

0.22.3    2013-08-08 14:52:28
          - Updated toolchain
          - Fixed POD errors
          - Mooified

0.21.1    2013-05-21 13:30:36
          - Ignore plugin in Git
          - Moved Debian plugin to own dist

0.20.4    2013-05-15 21:30:56
          - Bumped Class::Usul version
          - Renamed Class::Usul::Functions::say to emit
          - Added missing dependency
          - Bumped version for PAUSE indexer

0.1.5     2013-05-14 23:06:42
          - Added words to 07spelling.t
          - Removed redundant use strict
          - Fixed version numbers
          - Moved in modules from Class::Usul
          - Added project files to master branch



inc/Bob.pm
# @(#)Ident: Bob.pm 2013-08-08 10:24 pjf ;

package Bob;

use strict;
use warnings;
use inc::CPANTesting;

sub whimper { print {*STDOUT} $_[ 0 ]."\n"; exit 0 }

BEGIN { my $reason; $reason = CPANTesting::should_abort and whimper $reason; }

use version; our $VERSION = qv( '1.21' );

use File::Spec::Functions qw( catfile );
use Module::Build;

sub new {
   my ($class, $p) = @_; $p ||= {}; $p->{requires} ||= {}; __is_above_min( $p );

   my $module      = $p->{module} or whimper 'No module name';
   my $distname    = $module; $distname =~ s{ :: }{-}gmx;
   my $class_path  = catfile( 'lib', split m{ :: }mx, "${module}.pm" );

   return __get_build_class( $p )->new
      ( add_to_cleanup     => __get_cleanup_list( $p, $distname ),
        build_requires     => $p->{build_requires},
        configure_requires => $p->{configure_requires},
        create_license     => 1,
        create_packlist    => 0,
        create_readme      => 1,
        dist_version_from  => $class_path,
        license            => $p->{license} || 'perl',
        meta_merge         => __get_resources( $p, $distname ),
        module_name        => $module,
        no_index           => __get_no_index( $p ),
        notes              => __get_notes( $p ),
        recommends         => $p->{recommends},
        requires           => $p->{requires},
        sign               => defined $p->{sign} ? $p->{sign} : 1,
        share_dir          => __get_share_dir( $p ), );
}

# Private functions
sub __is_above_min {
   my $p        = shift;
   my $perl_ver = $p->{_min_perl_ver} = $p->{requires}->{perl} || 5.008;

   CPANTesting::is_testing() and $] < $perl_ver and return;

   $] < $perl_ver and whimper "Minimum required Perl version is ${perl_ver}";

   return;
}

sub __is_src { # Is this the developer authoring a module?
   return -f 'MANIFEST.SKIP';
}

sub __get_build_class { # Which subclass of M::B should we create?
   my $p = shift; exists $p->{build_class} and return $p->{build_class};

   my $path = catfile( qw( inc SubClass.pm ) );

   -f $path or return 'Module::Build';

   open my $fh, '<', $path or whimper "File ${path} cannot open: ${!}";

   my $code = do { local $/ = undef; <$fh> }; close $fh;

   return Module::Build->subclass( code => $code );
}

sub __get_cleanup_list {
   my $p = shift; my $distname = shift;

   return [ 'Debian_CPANTS.txt', 'MANIFEST.bak', "${distname}-*",
            map { ( '*/' x $_ ).'*~' } 0 .. 5 ];
}

sub __get_git_repository {
   return (map  { s{ : }{/}mx; s{ @ }{://}mx; $_ }
           grep { m{ \A git }mx }
           map  { s{ \s+ }{ }gmx; (split ' ', $_)[ 1 ] }
           grep { m{ \A origin }mx }
              qx{ git remote -v 2>/dev/null })[ 0 ];
}

sub __get_no_index {
   my $p = shift;

   return { directory => $p->{no_index_dir} || [ qw( examples inc share t ) ] };
}

sub __get_notes {
   my $p = shift; my $notes = exists $p->{notes} ? $p->{notes} : {};

   # Optionally create README.md and / or README.pod files
   $notes->{create_readme_md } = defined $p->{create_readme_md}
                               ? $p->{create_readme_md } :  1;
   $notes->{create_readme_pod} = $p->{create_readme_pod} || 0;
   $notes->{is_cpan_testing  } = CPANTesting::is_testing();
   # Add a note to stop CPAN testing if requested in Build.PL
   $notes->{stop_tests       } = CPANTesting::test_exceptions( $p );
   $notes->{url_prefix       } = defined $p->{url_prefix} ? $p->{url_prefix}
                               : 'https://metacpan.org/module/';
   $notes->{version          } = $VERSION;
   return $notes;
}

sub __get_repository { # Accessor for the VCS repository information
   my $repo;

   -d q(.git) and $repo = __get_git_repository() and return $repo;
   -d q(.svn) and $repo = __get_svn_repository() and return $repo;

   return;
}

sub __get_resources {
   my $p         = shift;
   my $distname  = shift;
   my $tracker   = defined $p->{bugtracker}
                 ? $p->{bugtracker}
                 : 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=';
   my $resources = $p->{resources} || {};

   $tracker and $resources->{bugtracker} = $tracker.$distname;
   $p->{home_page} and $resources->{homepage} = $p->{home_page};
   $resources->{license} ||= 'http://dev.perl.org/licenses/';

   # Only get repository info when authoring a distribution
   my $repo; __is_src and $repo = __get_repository
      and $resources->{repository} = $repo;

   return { resources => $resources };
}

sub __get_share_dir {
   my $p = shift; defined $p->{share_dir} and return $p->{share_dir};

   return -d 'share' ? 'share' : undef;
}

sub __get_svn_repository {
   return (grep { ! m{ \A file: }mx }
           (split q( ), (grep { m{ \A URL: }mx }
                            qx{ svn info })[ 0 ])[ 1 ])[ 0 ];
}

1;

# Local Variables:
# mode: perl
# tab-width: 3
# End:


Build.PL
# @(#)$Ident: Build.PL 2013-08-13 12:00 pjf ;

use strict;
use warnings;

use Module::Build;

sub whimper { print {*STDOUT} $_[ 0 ]."\n"; exit 0 }

$] >= 5.01 or $ENV{PERL5_CPANPLUS_IS_VERSION}
   or whimper "Minimum required Perl version is 5.01";

my $code = do { local $/; <DATA> };

Module::Build->subclass( code => $code )->new(
  'module_name'   => 'Class::Usul::Build',
  'notes'         => {
     is_reporting => !! ($ENV{AUTOMATED_TESTING} || $ENV{PERL_CR_SMOKER_CURRENT}
                         || ($ENV{PERL5OPT} || q()) =~ m{ CPAN-Reporter }mx),
     version      => 0.18.1 },
  'share_dir'     => -d 'share' ? 'share' : undef,
  'build_requires' => {
    'Module::Build' => '0.4004',
    'version' => '0.88'
  },
  'configure_requires' => {
    'Module::Build' => '0.4004',
    'version' => '0.88'
  },
  'requires' => {
    'Class::Usul' => 'v0.23.0',
    'MRO::Compat' => '0.11',
    'Module::Metadata' => '1.000011',
    'Moo' => '1.003000',
    'Perl::Version' => '1.011',
    'Try::Tiny' => '0.09',
    'local::lib' => '1.008004',
    'namespace::sweep' => '0.006',
    'parent' => '0.224'
  },
)->create_build_script;

exit 0;

__DATA__

sub _normalize_prereqs { # Dynamic config
   my $self = shift; my $p = $self->SUPER::_normalize_prereqs;

   my $osname = lc $^O;

   return $p;
}
